morf-core/src/main/java/org/alfasoftware/morf/sql/SqlTokenizer.java

```diff
@@ -0,0 +1,707 @@
+   *
+   * See https://www.postgresql.org/docs/8.0/sql-syntax.html#SQL-SYNTAX-COMMENTS
+   * See https://docs.oracle.com/en/database/oracle/oracle-database/19/sqlrf/Comments.html
+   *
+   * @param databaseType ORACLE/PGSQL
+   * @param query Query to be tokenized.
+   * @param length Total length of the query.
+   * @param i Position where this token matching should start.
+   * @param commentUnfinished This is set to true, when the comment is left unfinished before the query ends.
+   * @return End position of the token, if matched at this position, or negative number otherwise.
+   */
+  private static int matchMultiLineComment(String databaseType, String query, int length, int i, AtomicBoolean commentUnfinished) {
+    boolean allowCommentNesting = false;
+
+    if (PGSQL.equals(databaseType)) {
+      allowCommentNesting = true;
+    }
+
+    if (query.charAt(i) == '/' && i+1 < length && query.charAt(i+1) == '*') {
+      i += 2;
+      while (i < length) {
+        // end of the multi-line comment
+        if (query.charAt(i) == '*' && i+1 < length && query.charAt(i+1) == '/')
+          return i + 2;
+
+        if (allowCommentNesting) {
+          if (query.charAt(i) == '/' && i+1 < length && query.charAt(i+1) == '*') {
+            int mlc = matchMultiLineComment(databaseType, query, length, i, commentUnfinished);
+            if (mlc > 0)
+              i = mlc;
+          }
+        }
+
+        i++;
+      }
+
+      commentUnfinished.set(true);
+      return -1;
+    }
+
+    return -1;
+  }
+
+
+  /**
+   * See https://www.postgresql.org/docs/current/sql-syntax-lexical.html#SQL-SYNTAX-IDENTIFIERS
+   * See https://docs.oracle.com/en/database/oracle/oracle-database/19/sqlrf/Database-Object-Names-and-Qualifiers.html
+   */
+  private static boolean isIdentifierStartCharacter(char c) {
+    return 'A' <= c && c <= 'Z'
+        || 'a' <= c && c <= 'z'
+        || c == '_';
+  }
+
+
+  /**
+   * See https://www.postgresql.org/docs/current/sql-syntax-lexical.html#SQL-SYNTAX-IDENTIFIERS
+   * See https://docs.oracle.com/en/database/oracle/oracle-database/19/sqlrf/Database-Object-Names-and-Qualifiers.html
+   */
+  private static boolean isIdentifierMidCharacter(char c) {
+    return isIdentifierStartCharacter(c)
+        || '0' <= c && c <= '9'
+      //|| c == '#' // Oracle allows # in identifiers
+        || c == '$';
+  }
+
+
+  /**
+   * See https://www.postgresql.org/docs/current/sql-syntax-lexical.html#SQL-SYNTAX-CONSTANTS-NUMERIC
+   * See https://docs.oracle.com/en/database/oracle/oracle-database/19/sqlrf/Literals.html
+   */
+  private static boolean isNumberStartCharacter(char c) {
+    return '0' <= c && c <= '9';
+  }
+
+
+  /**
+   * See https://www.postgresql.org/docs/current/sql-syntax-lexical.html#SQL-SYNTAX-CONSTANTS-NUMERIC
+   * See https://docs.oracle.com/en/database/oracle/oracle-database/19/sqlrf/Literals.html
+   */
+  private static boolean isNumberMidCharacter(char c) {
```

Comment:
is this method necessary?  Could you not just have a single isNumberCharacter() method?

Reply:
Depends on whether we will want to start supporting floating numbers like "5.8" in the future.
I had this method in because I thought I would try to read the full numbers, including decimal points, etc.
But then it seemed completely unnecessary for our current purposes, and when reading vendor documentations, dialects seem to differ ever so slightly, which solidified my decision to skip parsing numbers in full detail. Rhis method remained as a remnant of my attempt. It could be useful when we decide to support full numbers in the future. In the meantime, does not do much harm.